otherlibs/systhreads/st_stubs.c

```diff
@@ -420,11 +419,15 @@ CAMLprim value caml_thread_initialize(value unit)   /* ML */
 {
   CAMLparam0();
 
+  if (!caml_domain_alone())
```

Comment:
Good catch.

Reply:
I wondered the same; are we sure that we control initialization order enough or is there a reason one should want to initialize Threads while other domains are running? In any case, this PR goes for the simplest solution. The more general problem might require a much more complicated solution and it is best addressed separately. Should an issue be opened for this?

Reply:
I think there's something generally worrying if a domain were to be spawned before we get to this codepath, because the domain in question would not go through `caml_thread_initialize_domain` (since we couldn't register the callback with `Domain.at_startup`, as done in `thread.ml`.).

I don't think `systhreads` should be setting up in the presence of Multiple domains, and the current code (as of trunk) is written under the assumption that it is done before any other domains is spawned.
In that regard I think your solution fixes a nefarious scenario and we should go with it.

However we should definitely open a issue to discuss more this problem.
I see this being a problem with user-code also relying on `Domain.at_startup` and being unexpectedly hit with a top-level `Domain.spawn`. Maybe it is fine? Maybe we should add restrictions somewhere in the pipeline? (restricting top-level domain spawn?). Likely we should document and describe best-practices.

Reply:
Yes, the problematic case is a library that 1- is linked before the Threads library in linking order, and 2- whose initialization code spawns one or several domains.  1- is not very likely, as Threads is a "base" library that is used by many others, and 2- is very bad programming in my opinion.  So it's safer to have a dynamic check, knowing that it should never trigger in real use.

I'm afraid there is no way to prevent module initialization code to spawn domains, and it's a legitimate thing to do for the initialization code of the final module of a standalone application !  However, modules that we put in libraries tend to have well-behaved initialization code with no observable effects...

Reply:
@Engil Another thing I saw is that loading Thread overwrites the default hooks that deal with the backup thread. How is the backup thread woken up with Threads currently? E.g. imagining the case where all threads are in a blocking section and you need a backup thread to serve a STW request.

Maybe there is there an opportunity to simplify everything and have Thread automatically initialized (before any OCaml code runs). What use-case serves _not_ having (sys)Threads?

Anyway let's split this discussion from this PR.

Reply:
@gadmm we can discuss about this in more details in another issue (or someplace else), the gist of it though is that if you look in `st_posix.h`, you will find that the `st_masterlock_release`/`st_masterlock_acquire` cycle handles the duty of messaging or not the backup thread depending on whether we have waiters or not.
There some comments around the functions dealing with the release/acquire cycle, feel free to page me if you need more details.

Reply:
Alright, I missed that. Looks like what I expected, thanks.

Reply:
> I don't see anything conceptually wrong with library initialization code that would, say, decide build an immutable representation of /usr/share/zoneinfo at startup, traversed and processed in parallel (so: spawn several domains using a parmap-style helper

One thing that is wrong here is the inability to handle errors: if something goes wrong while reading and preprocessing /usr//share/zoneinfo, an exception will be raised that no user code can handle.  That's one of the many reasons why library initialization code should perform very little computation, leaving heavy computation (e.g. that can fail or that can take significant time) to explicit `init` functions.